Solutions/Corvus.Azure.Cosmos.Tenancy/Corvus/Azure/Cosmos/Tenancy/RecreatableContainer.cs

```diff
@@ -0,0 +1,43 @@
+﻿// <copyright file="RecreatableContainer.cs" company="Endjin Limited">
+// Copyright (c) Endjin Limited. All rights reserved.
+// </copyright>
+
+namespace Corvus.Azure.Cosmos.Tenancy
+{
+    using System;
+    using System.Threading.Tasks;
+    using Microsoft.Azure.Cosmos;
+
+    /// <summary>
+    /// Represents a recreatable container.
+    /// </summary>
+    public struct RecreatableContainer
+    {
+        private readonly Func<Task<Container>> containerFactory;
+
+        /// <summary>
+        /// Initializes a new instance of the <see cref="RecreatableContainer"/> struct.
+        /// </summary>
+        /// <param name="containerFactory">The factory method to create a new instance of the container.</param>
+        /// <param name="instance">The initial container instance.</param>
+        public RecreatableContainer(Func<Task<Container>> containerFactory, Container instance)
+        {
+            this.containerFactory = containerFactory;
+            this.Instance = instance;
+        }
+
+        /// <summary>
+        /// Gets the <see cref="Container"/>.
+        /// </summary>
+        public Container Instance { get; private set; }
```

Comment:
Yikes! A mutable struct! 😨

Perhaps we could avoid that by having `RecreateContainer` below return a new `RecreatableContainer`?

But this has me wondering whether this is the correct model. Another way to do this would be to offer a variant on the existing method `ITenantCosmosContainerFactory`supplies for getting a container from a definition. Invoking this variant (specific to `IRecreatableTenantCosmosContainerFactory`) would indicate that you believe that the existing container is no longer any good, so you don't want a cached one, you want a new one. (It would probably be a good idea for this method also to accept the now-broken `Container` as an argument so that it can detect when the caller isn't the first one to notice this problem, and the relevant container has in fact already been taken out of circulation. This would avoid a potential problem on busy servers where you might suddenly get a large number of operations simultaneously noticing that the container is no longer any good, and they all ask for a new one. You'd only kick off the creation of a new container if the one passed in is the very same one you would otherwise have handed out.)

The common characteristic this alternate approach shares with the approach currently shown is that you've got a method you can call if you have reason to believe that the container you've got is no longer any good, and needs to be recreated. The difference is that we wouldn't need to have this `RecreatableContainer` type. You'd just invoke the method on the factory instead. So that's one less moving part, with no obvious downside, I think.

It also occurs to me that "this one has stopped working due to key rotation" and "we've had a config change, so we know some extant containers are no longer any good" are actually distinct scenarios requiring different handling. The thing about the former is that the problem's going to be unmissable: if key rotation stopped your container from working you're not going to be able to plough on regardless. However, if a config change means your container is no longer correctly configured, it might be possible to carry on without realising this because you don't notice anything obviously wrong. (Maybe. I'm not sure because I don't really know what these config changes might look like.) For the first one you want some way of saying "I've got this container but it's stopped working, so I need a new one, and you'd better not hand me back the same one." But for config changes you want some way of saying "If there is a container currently cached for this particular definition, it's no longer up to date, and you should stop handing it out." (And without adding a wrapper of some sort, there's no way to revoke ones already handed out.)

The distinction there is that in the case where you've got some config watcher that notices when relevant settings have changed, it doesn't know which, if any, containers are affected. It just knows that _if_ there's a cached container, it needs to be removed from the cache.

In short, one scenario is reactive, and the other pre-emptive. One size might not fit both particularly well.

Reply:
It is certainly the case that this deliberately doesn't contain any policy as to *why* you might want to create the container. Just the mechanism that does so.

I also agree about the mutability of this struct; that's horrible. I don't know what I was thinking.